internal/cloud/backend.go

```diff
@@ -204,21 +204,25 @@ func (b *Cloud) PrepareConfig(obj cty.Value) (cty.Value, tfdiags.Diagnostics) {
 	return obj, diags
 }
 
-// configureGenericHostname aliases the cloud backend hostname configuration
-// as a generic "localterraform.com" hostname. This was originally added as a
+// Alias aliases the cloud backend hostname configuration
+// as a generic hostname. This was originally added as a
 // Terraform Enterprise feature and is useful for re-using whatever the
 // Cloud/Enterprise backend host is in nested module sources in order
 // to prevent code churn when re-using config between multiple
 // Terraform Enterprise environments.
-func (b *Cloud) configureGenericHostname() {
+func (b *Cloud) Alias(aliasHostname string) error {
 	// This won't be an error for the given constant value
-	genericHost, _ := svchost.ForComparison(genericHostname)
+	genericHost, err := svchost.ForComparison(aliasHostname)
+	if err != nil {
+		return fmt.Errorf("unusable alias hostname: %w", err)
+	}
 
 	// This won't be an error because, by this time, the hostname has been parsed and
 	// service discovery requests made against it.
 	targetHost, _ := svchost.ForComparison(b.hostname)
 
 	b.services.Alias(genericHost, targetHost)
```

Comment:
This does seem overall reasonable but I worry a little about this implicit assumption that `b.services` here is always the same object as the service discovery object used by the module and provider installers in `terraform init`. I imagine that's true today but it seems relatively easy for future changes to break that assumption if e.g. we end up needing to give different parts of Terraform different views of the discovery information for some reason yet to be determined.

While I'm not sure how likely that actually is, what bothers me is that someone could potentially do it without realizing that they are breaking a non-obvious assumption. For everything else about the service discovery system we treat it as something that's set once inside `package main` and then not mutated again, and so it would be reasonable for a new contributor to not notice this special exception to that.

I'm honestly not really sure what to suggest instead, though. :confounded: One idea that came to my mind was to make this backend-level function just ask the question "what hostname should `localterraform.com` be an alias for?" and have it _return_ `targetHost` instead of directly write it into its own `b.services`, and then it could be the caller's responsibility to actually register it with the service discovery object. I think that would also make this function infallible (no `error` cases) because the only possible error here is that the given hostname is invalid, and under this new design this method would not actually need to know the alias hostname at all.

That is admittedly only a marginal improvement since it still assumes that everyone's sharing a single disco object, but I imagine the caller that would be doing it would be `command.Meta` and that's already a sort of "central controller" for everything else and so I have a hunch that it's more likely to be able to ensure that everyone's `svchost.Disco` gets this information, even if in future it ends up having to register this against more than one discovery object for some reason.

Reply:
@apparentlymart I tried my best to minimize the impact here, but it did turn out to be a marginal improvement because command.Meta is still modifying the service disco it doesn't own. But at least it's not happening one level further down in the backend. There was just no nice way I could see getting this alias information to the main package since the backend initialization is brokered by backendInit.